File: solutions/224.Basic_Calculator/AC_stack_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_stack_n.cpp
*  Create Date: 2015-09-07 19:07:22
*  Descripton:   
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    int calculate(string s) {
        stack<int> nums;
        stack<int> signs;
        int num = 0, res = 0, sign = 1;
        for (auto ch : s) {
            if (ch == ' ') continue;
            if (isdigit(ch)) {
                num = num * 10 + (ch - '0');
            } else {
                res += num * sign;
                num = 0;
                if (ch == '+') sign = 1;
                if (ch == '-') sign = -1;
                if (ch == '(') {
                    nums.push(res);
                    signs.push(sign);
                    res = 0;
                    sign = 1;
                }
                if (ch == ')') {
                    res = nums.top() + res * signs.top();
                    nums.pop();
                    signs.pop();
                }
            }
        }
        res += sign * num;
        return res;
    }
};

int main() {

    return 0;
}

File: solutions/224.Basic_Calculator/AC_stack_n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_stack_n.java
*  Create Date: 2015-09-07 20:36:56
*  Descripton:   
*/


import java.util.*;

class Main {
    public static void main(String[] args){
        System.out.println(Solution.calculate("(7)-(0)+(4)"));
    }
}

class Solution {
    public static int calculate(String s) {
        int res = 0, num = 0, sign = 1, bsign = 1;
        Stack<Integer> signs = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == ' ') continue;
            if (c >= '0' && c <= '9') {
                num = num * 10 + (c - '0');
            } else {
                res += num * sign * bsign;
                num = 0;
                if (c == '+') {
                    sign = 1;
                }
                if (c == '-') {
                    sign = -1;
                }
                if (c == '(') {
                    bsign *= sign;
                    signs.push(sign);
                    sign = 1;
                }
                if (c == ')') {
                    bsign *= signs.peek();
                    signs.pop();
                }
            }
        }
        res += num * sign * bsign;
        return res;
    }
}

